kv/kvserver: TestRaftPreVote failed

kv/kvserver.TestRaftPreVote [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11528135?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11528135?buildTab=artifacts#/) on master @ [6a311204b58cdd586e1951c9fa50c4d0ecd31273](https://github.com/cockroachdb/cockroach/commits/6a311204b58cdd586e1951c9fa50c4d0ecd31273):


```
=== RUN   TestRaftPreVote
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/ec76f3007984602f26e3543a773d47f5/logTestRaftPreVote1331143345
    test_log_scope.go:81: use -show-logs to present logs inline
=== CONT  TestRaftPreVote
    client_raft_test.go:6304: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/ec76f3007984602f26e3543a773d47f5/logTestRaftPreVote1331143345
--- FAIL: TestRaftPreVote (43.24s)
=== RUN   TestRaftPreVote/partial=true/quiesce=false/expLease=false
    test_server_shim.go:121: automatically injected virtual cluster under test; see comment at top of test_server_shim.go for details.
        (override via TestingSetDefaultTenantSelectionOverride)
    client_raft_test.go:6209: n1 has lease
    client_raft_test.go:6217: n3 not quiesced
    client_raft_test.go:6222: n3 partitioned
    client_raft_test.go:6227: n1 StateLeader at term=6 commit=22
    client_raft_test.go:6235: n3 not quiesced
    client_raft_test.go:6241: n3 StateFollower at term=6 commit=22
    client_raft_test.go:6139: r65 write rejected: GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00"), GC [/Table/Max,"\xfa\x00")
    client_raft_test.go:6241: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6244: n3 became pre-candidate
    client_raft_test.go:6251: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6251: n3 StatePreCandidate at term=6 commit=22
    client_raft_test.go:6254: n3 is still pre-candidate
    client_raft_test.go:6261: n1 StateLeader at term=6 commit=23
    client_raft_test.go:6265: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_raft_test.go:6265
        	            				github.com/cockroachdb/cockroach/pkg/testutils/subtest.go:23
        	Error:      	Not equal: 
        	            	expected: 0x16
        	            	actual  : 0x17
        	Test:       	TestRaftPreVote/partial=true/quiesce=false/expLease=false
            --- FAIL: TestRaftPreVote/partial=true/quiesce=false/expLease=false (5.79s)
=== RUN   TestRaftPreVote/partial=true
    --- FAIL: TestRaftPreVote/partial=true (21.28s)
=== RUN   TestRaftPreVote/partial=true/quiesce=false
        --- FAIL: TestRaftPreVote/partial=true/quiesce=false (12.63s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRaftPreVote.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31022

--- a/pkg/kv/kvserver/client_raft_test.go
+++ b/pkg/kv/kvserver/client_raft_test.go
@@ -6127,16 +6127,14 @@ func TestRaftPreVote(t *testing.T) {
 
 				// We don't want any writes to the range, to avoid the follower from
 				// falling behind on the log and failing prevotes only because of that.
-				// We install a request filter which rejects writes to the range during
-				// the partition (typically txn record cleanup via GC requests).
-				//
-				// We reject requests, not proposals, because we don't want to mess too
-				// much with the Raft machinery and mask other issues.
+				// We install a proposal filter which rejects proposals to the range
+				// during the partition (typically txn record cleanup via GC requests,
+				// but also e.g. lease extensions).
 				var partitioned atomic.Bool
 				var rangeID roachpb.RangeID
-				reqFilter := func(ctx context.Context, ba *kvpb.BatchRequest) *kvpb.Error {
-					if partitioned.Load() && ba.IsWrite() && ba.RangeID == rangeID {
-						t.Logf("r%d write rejected: %s", rangeID, ba)
+				propFilter := func(args kvserverbase.ProposalFilterArgs) *kvpb.Error {
+					if partitioned.Load() && args.Req.RangeID == rangeID {
+						t.Logf("r%d proposal rejected: %s", rangeID, args.Req)
 						return kvpb.NewError(errors.New("rejected"))
 					}
 					return nil
@@ -6160,7 +6158,7 @@ func TestRaftPreVote(t *testing.T) {
 						Knobs: base.TestingKnobs{
 							Store: &kvserver.StoreTestingKnobs{
 								DisableQuiescence:            !quiesce,
-								TestingRequestFilter:         reqFilter,
+								TestingProposalFilter:        propFilter,
 								DisableAutomaticLeaseRenewal: true,
 							},
 						},

